### DISCUSSION TITLE
Help with querying for the binary's version at runtime

### DISCUSSION THREAD

**Discussion by dlvhdr** [goreleaser/goreleaser#3780] [Q&A] [answered]
Hi friends :)

I'm adding a `--version` flag to my CLI app, and trying to use goreleaser for it.
I tried following [this doc](https://goreleaser.com/cookbooks/build-go-modules/?h=proxy) but failed.

I then tried following what goreleaser itself does by providing flags to the compiler with:
```yaml
ldflags:
    - -s -w -X main.version={{.Version}} -X main.commit={{.Commit}} -X main.date={{ .CommitDate }} -X main.builtBy=goreleaser
```
And then accessing them with `debug.ReadBuildInfo()`.

When running `goreleaser build --clean` I faced the following errors:
1. It seems I have to push the tag to origin for the local build to work
2. I am now facing the error: `error=failed to proxy module: exit status 1: go: github.com/dlvhdr/gh-dash@v3.7.0-next: invalid version: go.mod has post-v3 module path "github.com/dlvhdr/gh-dash/v3" at revision v3.7.0-next`

My code can be found on this branch in my repo [dlvhdr/gh-dash@v3.7.0-next](https://github.com/dlvhdr/gh-dash/tree/v3.7.0-next)

Appreciate any help!

**Comment by caarlos0** [comment, marked as answer]
Hey!

OK, a couple of things:

1. these need to be public fields https://github.com/dlvhdr/gh-dash/blob/2b44bd659d8f289dd48f7ac7f523a3ddf4c464bf/cmd/root.go#L22-L25
2. the `-X main.version` flags need to point to that cmd pkg, you can try it out locally with `go build` to figure out the right combination
3. if you set `gomod.enabled`, yes, it'll require the tag to be pushed, because it'll use the code from the go mod proxy
4. 3 is also why you get your second error: you pushed the tag, build it (which got it proxied to the go mod proxy), deleted it, changed something, retagged, pushed again, and tried to build again, and now the tag in the go mod proxy does not match the tag in git... tl;dr: do not recreate tags with go mod proxying enabled.

hope that clear things up and helps a little bit :)

**Reply by dlvhdr** [reply]
Thanks @caarlos0! This seems to work.
I didn't know about `ldflags` and especially how the `-X` flag works.
In addition I'm not quite sure I understand what `gomod.proxy` does, but after disabling it, I seem to be able to run goreleaser in a Github Action. I got a similar error as before - something about go expecting my `go.mod` file to contain the version as a suffix (`/v3`)
Would you like me to add some links to info about this in the goreleaser docs?

**Reply by caarlos0** [reply]
go mod proxy emulates building as `go install` would, verifying deps against the go module proxy.

if you do it one time for a given tag, you shouldn't recreate it, ever. It'll cause errors for other people when running `go get` and `go install` otherwise.

the `/v3` suffix shouldn't make a diference afaik.

and yeap, feel free to add to the docs where you feel it makes sense or if you think something is currently lacking. Thanks!
